Add `itself`, `not` and `iif` to conditional tranformations

nepalez opened the pull request:
`itself` returns the unchanged value
```ruby
  fn = t(:itself)
  fn[:foo] # => :foo
```
`not` negates the result of another transformation
```ruby
  fn = t(:not, -> value { value.is_a? String })
  fn[:foo]  # => true
  fn['foo'] # => false
```
`iif` runs a transformation depending on predicate
```ruby
  fn = t(:iif, -> value { value['o'] }, -> value { value.upcase }, -> value { value.downcase })
  fn['Foo'] # => 'FOO'
  fn['Bar'] # => 'bar'
```
Also refactored `guard` and `is` using the existing transprocs, added specs for the `is`.

solnic commented:
This seems useful. Only wondering about `iif` name, is this a convention taken from somewhere? I've never seen it

nepalez commented:
This is from VB http://www.techonthenet.com/access/functions/advanced/iif.php

adz commented:
I've used "IIF" in office apps...

I noted from here ( https://en.wikipedia.org/wiki/IIf ) is that IIF is an "abbreviation for Inline if", and

    Another issue with IIf arises because it is a library function: 
    unlike the C-derived conditional operator, both truepart and the falsepart 
    will be evaluated regardless of which one is actually returned. 

That doesn't seem to match semantically, right?

Maybe it should match up with Lisp semantics... just use :if but potentially also :cond 
http://docs.racket-lang.org/reference/if.html

solnic commented:
Yes I think this may confuse people

solnic commented:
How about `:if_true` and `:if_false`? Would be super explicit

nepalez commented:
I thing the only reason for this `iif` is it defines both branches at once, like in ruby
```ruby
condition ? if_true : if_false
```
Otherwise we already have `is` and it does the work. I don't think we need another one.

Concerning `itself`... now I'd rename it to `identity` and implement it slightly different with default `nil`:
```ruby
def self.identity(value = nil)
  value
end
```

More generally, recently I extracted the [selector](https://github.com/nepalez/selector) from a `faceter`, and it implements conditions via unified API. That's why I in doubt. Maybe we don't need `not` as well?

As a result today I'd leave only `identity` (ex-`itself`) from the whole PR. (tbh I've completely forgotten about it).

AMHOL commented:
How about `ternary` then?

nepalez commented:
@AMHOL why not?

But personally I'd prefer to postpone it to the moment `transproc` is splitted to core/libs
(this PR was sent before local `Registry`, that changed my own approach)

nepalez commented:
For now i updated the PR:
* removed `iif`
* moved `itself` to `Coercions[:identity]` (to follow category notation for `Id`entity arrow).
* updated `not` following the new syntax (for `Registry`)
